wavesexchange_repos/src/circuit_breaker/mod.rs

```diff
@@ -0,0 +1,299 @@
+        self
+    }
+
+    /// Build the circuit breaker.
+    /// Note: you should set up all 3 fields within this builder, either it will panic
+    pub fn build(self) -> Result<CircuitBreaker<S>, S::Error> {
+        // probably there is a better way to force use all with_* methods on builder
+        if self.max_err_count_per_timespan.is_none() {
+            panic!("max_err_count_per_timespan is not set");
+        }
+
+        if self.max_timespan.is_none() {
+            panic!("max_timespan is not set");
+        }
+
+        if self.init_fn.is_none() {
+            panic!("init_fn is not set");
+        }
+
+        let init_fn = self.init_fn.unwrap();
+
+        Ok(CircuitBreaker {
+            state: RwLock::new(CBState {
+                data_source: Arc::new(init_fn()?),
+                err_count: 0,
+                first_err_ts: None,
+            }),
+            max_timespan: self.max_timespan.unwrap(),
+            max_err_count_per_timespan: self.max_err_count_per_timespan.unwrap(),
+            init_fn,
+        })
+    }
+}
+
+impl<S: FallibleDataSource> CircuitBreaker<S> {
+    pub fn builder() -> CircuitBreakerBuilder<S> {
+        CircuitBreakerBuilder::new()
+    }
+
+    pub fn builder_from_cfg(cfg: &Config) -> CircuitBreakerBuilder<S> {
+        Self::builder()
+            .with_max_err_count_per_timespan(cfg.max_err_count_per_timespan)
+            .with_max_timespan(cfg.max_timespan)
+    }
+
+    /// Query the data source. If succeeded, CB resets internal error counter.
+    /// If error returned, counter increases.
+    /// If (N > max_err_count_per_timespan) errors appeared, CB is falling back (panic as default).
+    /// If not enough errors in a timespan appeared to trigger CB's fallback, error counter will be reset.
+    pub async fn query<T, F, Fut>(&self, query_fn: F) -> Result<T, S::Error>
+    where
+        F: FnOnce(Arc<S>) -> Fut,
+        Fut: Future<Output = Result<T, S::Error>>,
+    {
+        let state_read_lock = self.state.read().await;
+        let result = query_fn(state_read_lock.data_source.clone()).await;
+        let old_err_count = state_read_lock.err_count;
+
+        drop(state_read_lock);
+
+        if let Err(e) = &result {
+            if S::is_countable_err(e) {
+                let mut state = self.state.write().await;
+                state.inc();
+
+                debug!("err count: {}", state.err_count);
+
+                match state.first_err_ts {
+                    Some(ts) => {
+                        let elapsed = ts.elapsed();
+
+                        if state.err_count <= self.max_err_count_per_timespan {
+                            if elapsed > self.max_timespan {
+                                state.reset();
+                            }
+                        } else {
+                            return Err(state
+                                .data_source
+                                .fallback(elapsed.as_millis(), state.err_count));
```

Comment:
What is the use of that `fallback()` function (name is unclear), why not to just make `query()` return a dedicated error type `enum QueryError<E> { Error(E), CircuitBreakerTriggered { elapsed: Duration, err_count: u32 } }` and let the application handle errors normally? Very strange design to me.